godotproject/Menus/HighScoreScreen/HighScoreScreen.gd

```diff
@@ -0,0 +1,238 @@
+		return
+	# Save to disk
+	_save_to_disk()
+	# Set up everything required to animate the score coming in.
+	display_level_data.append(dict)
+	current_idx = 5
+	_display_score_data(display_level_data, current_idx)
+	_change_state(State.INITIAL_DISPLAY)
+
+func _change_state(state):
+	# If the state has a timer, start it.
+	if state == State.INITIAL_DISPLAY:
+		$InitialTimer.start()
+	elif state == State.THONK:
+		$ThonkTimer.start()
+	elif state == State.FINAL_SOUND:
+		$FinalSoundTimer.start()
+	current_state = state
+
+func _step_thonk_animation():
+	var current = display_level_data[current_idx]
+	display_level_data.remove(current_idx)
+	current_idx -= 1
+	display_level_data.insert(current_idx, current)
+	if current_idx != 5 and len(display_level_data) > 5:
+		display_level_data.pop_back()
+	_display_score_data(display_level_data, current_idx)
+	# This looks kinda funky, but it'll play the following notes:
+	# C D E G c
+	# It should sound nicer the higher of a place you get.
+	# Get 1st place to try it out :)
+	var rate = [2,3.0/2,5.0/4,9.0/8,1][current_idx]
+	$ThonkSmallSFX.stop()
+	$ThonkSmallSFX.pitch_scale = rate
+	$ThonkSmallSFX.play()
+	if current_idx <= final_idx:
+		_change_state(State.FINAL_SOUND)
+
+func _play_final_sound():
+	# Don't play sound if we didn't get on the scoreboard
+	if final_idx != -1:
+		# Play sound of score settling into place
+		$ThonkBigSFX.play()
+	_change_state(State.WAIT_FOR_PRESS)
+
+# This function is unit testable
+func _compare_scores(new, old):
+	return new['stars'] > old['stars'] or (new['stars'] == old['stars'] and new['time'] < old['time'])
+
+# This function is unit testable
+func _add_score_to_current_level(dict):
+	var target_i = -1
+	for i in range(len(current_level_data)):
+		var test_i = len(current_level_data)-1-i
+		var test_dict = current_level_data[test_i]
+		if _compare_scores(dict, test_dict):
+			target_i = test_i
+		else:
+			break
+	if target_i != -1:
+		current_level_data.insert(target_i, dict)
+		current_level_data.pop_back()
+	return target_i
+
+# This function is unit testable
+func _calculate_score(stars, time):
+	var time_inv = max(150 - time, 0)
+	return stars * 1000 + (floor(time_inv * 60) * 1000 / 1800)
+
+func _change_level(name):
+	if current_level and current_level_data:
+		# Make sure we've saved our current level data to disk.
+		# (In case we updated it and didn't persist it before).
+		# We need to do this now because this will properly deal with the
+		# "current_level_data".
+		_save_to_disk()
+	current_level = name
+	if name in all_level_data:
+		current_level_data = all_level_data[name]
+	else:
+		# We have to create a new set of level data for this level.
+		current_level_data = []
+		for i in range(5):
+			# Make some fake high scores
+			var stars = max(0, 2-i)
+			stars *= stars
+			stars = min(3, stars)
+			var time = 30*i+30
+			var score = _calculate_score(stars,time)
+			current_level_data.append({'stars':stars, 'time':time, 'score':score})
```

Comment:
Generation of high scores when there wasn't already data

Reply:
Instead of this being random highscore can the entries all just be 0? That way they will automatically get pushed out